### PR TITLE
Add all local users to the user_directory and optionally search them

### PR DESCRIPTION
**Pull request by ara4n**
Quick and dirty fix to https://github.com/matrix-org/synapse/issues/2720.

### CODE REVIEW

**Comment by ara4n** [comment]
@erikjohnston ptal. i'm particularly baffled on how the FTS match could ever have worked on sqlite - I see no evidence that `(foo* | foo)` is ever valid as a match term?

**Comment by ara4n** [comment]
@erikjohnston one design thought on this: i have a feeling that the `include_pattern` idea is overkill (as it causes constant filtering based on the LIKE pattern) and it should just be a `search_all_local_users` flag. wdyt?

**Comment by erikjohnston** [comment]
> @erikjohnston one design thought on this: i have a feeling that the include_pattern idea is overkill (as it causes constant filtering based on the LIKE pattern) and it should just be a search_all_local_users flag. wdyt?

Yeah, I think having a flag is going to be best here, I think using a LIKE expression may tie our hands if we need to change how the lookup works (if e.g. it turns out to be slow)

File: synapse/handlers/profile.py [hunk 1]
```diff
@@ -183,6 +191,12 @@ def set_avatar_url(self, target_user, requester, new_avatar_url, by_admin=False)
             target_user.localpart, new_avatar_url
         )
 
+        if self.hs.config.user_directory_include_pattern:
+            profile = yield self.store.get_profileinfo(target_user.localpart)
+            yield self.user_directory_handler.handle_local_profile_change(
+                target_user.user_id, profile
+            )
```
Comment (erikjohnston) [inline synapse/handlers/profile.py]: I would move the if and fetching of profile into `handle_local_profile_change` rather than duplicating it everywhere

File: synapse/handlers/user_directory.py [hunk 2]
```diff
@@ -148,16 +159,30 @@ def _do_initial_spam(self):
...
         for room_id in room_ids:
-            logger.info("Handling room %d/%d", num_processed_rooms, len(room_ids))
+            logger.info("Handling room %d/%d", num_processed_rooms+1, len(room_ids))
             yield self._handle_initial_room(room_id)
             num_processed_rooms += 1
             yield sleep(self.INITIAL_SLEEP_MS / 1000.)
 
         logger.info("Processed all rooms.")
 
+        if self.include_pattern:
```
Comment (erikjohnston) [inline synapse/handlers/user_directory.py]: In fact, I would be tempted to do this all the time and do the checks when we filter. Otherwise we need to keep track of people toggling the option on and off

File: synapse/handlers/user_directory.py [hunk 3]
```diff
@@ -148,16 +159,30 @@ def _do_initial_spam(self):
...
+            num_processed_users = 0
+            user_ids = yield self.store.get_all_local_users()
```
Comment (erikjohnston) [inline synapse/handlers/user_directory.py]: Do we want to include appservice users here?

Comment (erikjohnston) [reply synapse/handlers/user_directory.py]: I'm also worried that pulling out millions of users is going to be painful.

Comment (erikjohnston) [reply synapse/handlers/user_directory.py]: And its going to take matrix.org more than a day on the default settings to get through all its users

**Comment by erikjohnston** [comment]
Also, you broke the unit tests and flake8 is sad

File: synapse/handlers/user_directory.py (hunk 2 above)
Comment (ara4n) [reply synapse/handlers/user_directory.py]: i deliberately wasn't doing this (and instead thought the user can blow away the user_dir caches if they change the option), as for bigger servers it's going to make the user_directory and user_directory_search tables enormous, and slow down the FTS for a feature which may not even be being used...

File: synapse/handlers/user_directory.py (hunk 3 above)
Comment (ara4n) [reply synapse/handlers/user_directory.py]: hm, point. so your suggestion is to batch? and yes, we probably should be including AS users.

File: synapse/handlers/profile.py (hunk 1 above)
Comment (ara4n) [reply synapse/handlers/profile.py]: mm, point.

File: synapse/handlers/user_directory.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/handlers/user_directory.py]: Given that we add users to the table when they are in a room with anyone else, wouldn't the vast majority of users already be in there?

**Comment by ara4n** [comment]
@erikjohnston PTAL now.
 * changes the config option to be 'search all users', which certainly feels much more sane
 * fixes tests & pep8
 * speeds up the initial spam (but doesn't batch it; i think for the edge case of a server with millions of users, we can probably still temporarily slurp them all up into RAM)
 * spells out that if you change the option you'll need to flush the search tables

File: synapse/storage/profile.py [hunk 4]
```diff
@@ -26,6 +29,30 @@ def create_profile(self, user_localpart):
...
+    @defer.inlineCallbacks
+    def get_profileinfo(self, user_localpart):
+        try:
+            profile = yield self._simple_select_one(
+                table="profiles",
+                keyvalues={"user_id": user_localpart},
+                retcols=("displayname", "avatar_url"),
+                desc="get_profileinfo",
+            )
+        except StoreError, e:
```
Comment (erikjohnston) [inline synapse/storage/profile.py]: Should really be using `StoreError as e` syntax

**Comment by richvdh** [comment]
> Quick and dirty fix to #2720.

Just dirty, I'd say :/

**Comment by ara4n** [comment]
might need slightly more detailed review than that...
